# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [23]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [43]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [44]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [26]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [27]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [28]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [29]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [30]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [31]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [32]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [33]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [34]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [35]:
df['Season'].value_counts()

2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: Season, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [36]:
# Write your code here

# Assuming your data is in a list of tuples (year, value)
data = [
    (2016, 5369), (2014, 5362), (2015, 5354), (2013, 5320), 
    (2010, 5263), (2012, 5253), (2009, 5249), (2011, 5246), 
    (2008, 5163), (2007, 5043), (2006, 4757), (2005, 4675), 
    (2003, 4616), (2004, 4571), (2002, 4555), (2000, 4519), 
    (2001, 4467), (1999, 4222), (1998, 4167), (1997, 4155), 
    (1992, 4127), (1991, 4123), (1996, 4122), (1995, 4077), 
    (1994, 4060), # ... and so on for the rest of your data
    (1988, 3955), (1987, 3915), (1986, 3783), (1985, 3737)
]

# Creating a DataFrame
df2 = pd.DataFrame(data, columns=['Season', 'Value'])

# Using nunique() to count unique seasons
unique_seasons = df['Season'].nunique()

print(f"Number of unique seasons: {unique_seasons}")


Number of unique seasons: 29


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [37]:

import pandas as pd

# Example DataFrame
data = {
    'Season': [2016, 2014, 2015, ...],  # your season data
    'Wteam': ['Team A', 'Team B', 'Team A', ...]  # hypothetical winning team data
}

df = pd.DataFrame(data)

# Using value_counts() to count wins for each team
win_counts = df['Wteam'].value_counts()

# The team with the most wins
most_wins_team = win_counts.idxmax()
most_wins_count = win_counts.max()

print(f"Team with the most wins: {most_wins_team} ({most_wins_count} wins)")


Team with the most wins: Team A (2 wins)


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [41]:
df.iloc

In [46]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [47]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [48]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [49]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [50]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [51]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [52]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [53]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [54]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [55]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
49310     1997      66   1157      61   1204      21    H      0
89021     2006      44   1284      41   1343      21    A      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [62]:

# Sample data
data = {
    'Season': [100027, 49310, 89021, 85042, 103660],
    'Daynum': [2008, 1997, 2006, 2005, 2009],
    'Wteam': [66, 66, 44, 66, 26],
    'Wscore': [1203, 1157, 1284, 1131, 1326],
    'Lteam': [49, 61, 41, 73, 59],
    'Lscore': [1387, 1204, 1343, 1216, 1359],
    'Wloc': ['H', 'H', 'A', 'H', 'H'],
    'Numot': [0, 0, 0, 0, 0]
}

df = pd.DataFrame(data)

# Sorting by 'Season', 'Wteam', and 'Wscore'
df_sorted_season = df.sort_values(by='Season')
df_sorted_wteam = df.sort_values(by='Wteam')
df_sorted_wscore = df.sort_values(by='Wscore')

# Selecting rows and columns with iloc
# Note: Adjust the row indices according to your full dataset
selected_season = df_sorted_season.iloc[100:200][['Season', 'Wteam', 'Wscore']]
selected_wteam = df_sorted_wteam.iloc[100:200][['Season', 'Wteam', 'Wscore']]
selected_wscore = df_sorted_wscore.iloc[100:200][['Season', 'Wteam', 'Wscore']]

**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [63]:
# Write your code here

# Finding the game with the highest winning score
highest_score_game = df_sorted_wscore.iloc[0]

# Extracting the season and winning team for this game
season_highest_score = highest_score_game['Season']
winning_team_highest_score = highest_score_game['Wteam']

print(f"Season: {season_highest_score}, Winning Team: {winning_team_highest_score}")

Season: 85042, Winning Team: 66


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [65]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0  100027    2008     66    1203     49    1387    H      0
1   49310    1997     66    1157     61    1204    H      0
2   89021    2006     44    1284     41    1343    A      0
3   85042    2005     66    1131     73    1216    H      0
4  103660    2009     26    1326     59    1359    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [67]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Empty DataFrame
Columns: [Season, Daynum, Wteam, Wscore, Lteam, Lscore, Wloc, Numot]
Index: []

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [69]:


# Sample data
data = {
    'Season': [100027, 49310, 89021, 85042, 103660],
    'Daynum': [2008, 1997, 2006, 2005, 2009],
    'Wteam': [66, 66, 44, 66, 26],
    'Wscore': [1203, 1157, 1284, 1131, 1326],
    'Lteam': [49, 61, 41, 73, 59],
    'Lscore': [1387, 1204, 1343, 1216, 1359],
    'Wloc': ['H', 'H', 'A', 'H', 'H'],
    'Numot': [0, 0, 0, 0, 0]
}

df = pd.DataFrame(data)

# Creating 'ScoreDifference' column
df['ScoreDifference'] = (df['Wscore'] - df['Lscore']).abs()

# Calculating the average 'ScoreDifference'
average_score_difference = df['ScoreDifference'].mean()

# Filtering the DataFrame
filtered_df = df[df['ScoreDifference'] > average_score_difference]

print(filtered_df)

   Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference
0  100027    2008     66    1203     49    1387    H      0              184
3   85042    2005     66    1131     73    1216    H      0               85


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [70]:
# Write your code here

# Assuming 'filtered_df' is the DataFrame from the previous step

# Finding the game with the highest 'ScoreDifference'
highest_diff_game = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

# Extracting the season, winning team, and losing team for this game
season_highest_diff = highest_diff_game['Season']
winning_team_highest_diff = highest_diff_game['Wteam']
losing_team_highest_diff = highest_diff_game['Lteam']

print(f"Season: {season_highest_diff}, Winning Team: {winning_team_highest_diff}, Losing Team: {losing_team_highest_diff}")

Season: 100027, Winning Team: 66, Losing Team: 49


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [71]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
26    1326.000000
44    1284.000000
66    1163.666667
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [72]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
26     H       1
44     A       1
66     H       3
Name: Wloc, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [73]:
df.values

array([[100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
       [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
       [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
       [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
       [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [74]:
df.values[0][0]

100027

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [76]:
# Write your code here
import pandas as pd
import numpy as np


# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Season'] = df['Season'].astype(int)
df['Wscore'] = df['Wscore'].astype(int)

# Grouping by 'Season' and calculating the average 'Wscore'
average_winning_score_by_season = df.groupby('Season')['Wscore'].mean()

print(average_winning_score_by_season)

Season
49310     1157.0
85042     1131.0
89021     1284.0
100027    1203.0
103660    1326.0
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [77]:
# Write your code here

import pandas as pd
import numpy as np

# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Wteam'] = df['Wteam'].astype(int)
df['Wscore'] = df['Wscore'].astype(int)

# Grouping by 'Wteam' and calculating the maximum 'Wscore'
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()

print(max_winning_score_by_team)


Wteam
26    1326
44    1284
66    1203
Name: Wscore, dtype: int32


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [78]:
# Write your code here

import pandas as pd
import numpy as np

# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Season'] = df['Season'].astype(int)
df['Wteam'] = df['Wteam'].astype(int)
df['Wscore'] = df['Wscore'].astype(int)

# Grouping by 'Season' and 'Wteam' and calculating the average 'Wscore'
grouped_df = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# Finding the team with the highest average winning score for each season
highest_avg_score_by_season = grouped_df.reset_index().groupby('Season').apply(lambda x: x.loc[x['Wscore'].idxmax()])

print(highest_avg_score_by_season)

          Season  Wteam  Wscore
Season                         
49310    49310.0   66.0  1157.0
85042    85042.0   66.0  1131.0
89021    89021.0   44.0  1284.0
100027  100027.0   66.0  1203.0
103660  103660.0   26.0  1326.0


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [79]:
# Write your code here

import pandas as pd
import numpy as np

# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Season'] = df['Season'].astype(int)
df['Wteam'] = df['Wteam'].astype(int)

# Grouping by 'Season' and 'Wteam' and counting the number of wins
wins_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

print(wins_count)

   Season  Wteam  Wins
0   49310     66     1
1   85042     66     1
2   89021     44     1
3  100027     66     1
4  103660     26     1


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [81]:
# Write your code here

import pandas as pd
import numpy as np

# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Season'] = df['Season'].astype(int)
df['Wteam'] = df['Wteam'].astype(int)

# Grouping by 'Season' and 'Wteam' and counting the number of wins
wins_count = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')

# Finding the team with the most wins for each season
team_with_most_wins = wins_count.loc[wins_count.groupby('Season')['Wins'].idxmax()]

print(team_with_most_wins)

   Season  Wteam  Wins
0   49310     66     1
1   85042     66     1
2   89021     44     1
3  100027     66     1
4  103660     26     1


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [84]:
# Write your code here

import pandas as pd
import numpy as np

# Provided array
data_array = np.array([
    [100027, 2008, 66, 1203, 49, 1387, 'H', 0, 184],
    [49310, 1997, 66, 1157, 61, 1204, 'H', 0, 47],
    [89021, 2006, 44, 1284, 41, 1343, 'A', 0, 59],
    [85042, 2005, 66, 1131, 73, 1216, 'H', 0, 85],
    [103660, 2009, 26, 1326, 59, 1359, 'H', 0, 33]
], dtype=object)

# Creating DataFrame
df = pd.DataFrame(data_array, columns=['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot', 'ScoreDifference'])

# Converting the relevant columns to the appropriate data types
df['Lteam'] = df['Lteam'].astype(int)
df['Lscore'] = df['Lscore'].astype(int)
df['Wteam'] = df['Wteam'].astype(int)
df['Wscore'] = df['Wscore'].astype(int)

# Grouping by 'Lteam' and calculating the average 'Lscore'
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean()

# Grouping by 'Wteam' and calculating the average 'Wscore'
average_winning_score_by_team = df.groupby('Wteam')['Wscore'].mean()

# Comparing the average losing score with the average winning score
comparison_df = pd.DataFrame({
    'Avg_Losing_Score': average_losing_score_by_team,
    'Avg_Winning_Score': average_winning_score_by_team
})

print(comparison_df)

# 41, 49, 59, 61 and 73 have a higher average losing score than winning score

    Avg_Losing_Score  Avg_Winning_Score
26               NaN        1326.000000
41            1343.0                NaN
44               NaN        1284.000000
49            1387.0                NaN
59            1359.0                NaN
61            1204.0                NaN
66               NaN        1163.666667
73            1216.0                NaN


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [85]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             100027
Daynum               2008
Wteam                  66
Wscore               1203
Lteam                  49
Lscore               1387
Wloc                    H
Numot                   0
ScoreDifference       184
Name: 0, dtype: object
Season             49310
Daynum              1997
Wteam                 66
Wscore              1157
Lteam                 61
Lscore              1204
Wloc                   H
Numot                  0
ScoreDifference       47
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [87]:
# Write your code here

import pandas as pd

# Assuming you have a DataFrame named 'df'
# If the DataFrame is not named, replace 'df' with the actual name of your DataFrame

# Create a new column 'HighScoringGame'
df['HighScoringGame'] = 'No'  # Initialize with 'No' for all rows

for index, row in df.iterrows():
    if row['Wscore'] > 100:
        df.at[index, 'HighScoringGame'] = 'Yes'

# Display the updated DataFrame
print(df)

   Season Daynum  Wteam  Wscore  Lteam  Lscore Wloc Numot ScoreDifference  \
0  100027   2008     66    1203     49    1387    H     0             184   
1   49310   1997     66    1157     61    1204    H     0              47   
2   89021   2006     44    1284     41    1343    A     0              59   
3   85042   2005     66    1131     73    1216    H     0              85   
4  103660   2009     26    1326     59    1359    H     0              33   

  HighScoringGame  
0             Yes  
1             Yes  
2             Yes  
3             Yes  
4             Yes  


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [88]:
# Write your code here

import pandas as pd

# Assuming you have a DataFrame named 'df'
# If the DataFrame is not named, replace 'df' with the actual name of your DataFrame

# Initialize a dictionary to keep track of the number of games played by each team
team_games = {}

# Iterate through the rows of the DataFrame
for index, row in df.iterrows():
    # Update the dictionary for the winning team
    if row['Wteam'] in team_games:
        team_games[row['Wteam']] += 1
    else:
        team_games[row['Wteam']] = 1
    
    # Update the dictionary for the losing team
    if row['Lteam'] in team_games:
        team_games[row['Lteam']] += 1
    else:
        team_games[row['Lteam']] = 1

# Display the total number of games played by each team
for team, games in team_games.items():
    print(f"Team {team}: {games} games")

Team 66: 3 games
Team 49: 1 games
Team 61: 1 games
Team 44: 1 games
Team 41: 1 games
Team 73: 1 games
Team 26: 1 games
Team 59: 1 games


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [89]:
# Write your code here

import pandas as pd

# Assuming you have a DataFrame named 'df'
# If the DataFrame is not named, replace 'df' with the actual name of your DataFrame

# Initialize a dictionary to keep track of the highest score difference for each season
season_high_score_diff = {}

# Iterate through the rows of the DataFrame
for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['ScoreDifference']
    
    # Update the dictionary if the current game has a higher score difference for the season
    if season in season_high_score_diff:
        if score_diff > season_high_score_diff[season]['ScoreDifference']:
            season_high_score_diff[season] = {'GameID': index, 'ScoreDifference': score_diff}
    else:
        season_high_score_diff[season] = {'GameID': index, 'ScoreDifference': score_diff}

# Display the game with the highest score difference for each season
for season, info in season_high_score_diff.items():
    game_id = info['GameID']
    score_diff = info['ScoreDifference']
    print(f"Season {season}: Game ID {game_id} has the highest score difference of {score_diff}")

Season 100027: Game ID 0 has the highest score difference of 184
Season 49310: Game ID 1 has the highest score difference of 47
Season 89021: Game ID 2 has the highest score difference of 59
Season 85042: Game ID 3 has the highest score difference of 85
Season 103660: Game ID 4 has the highest score difference of 33


Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [92]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [91]:
# Write your code here

import pandas as pd

# Assuming you have a DataFrame named 'df'
# If the DataFrame is not named, replace 'df' with the actual name of your DataFrame

# Calculate the total number of games played by each team, whether they won or lost
winning_team_counts = df['Wteam'].value_counts()
losing_team_counts = df['Lteam'].value_counts()

# Add the two Series together to get the total number of games played by each team
total_games_by_team = winning_team_counts.add(losing_team_counts, fill_value=0)

# Display the total number of games played by each team
print(total_games_by_team)

26    1.0
41    1.0
44    1.0
49    1.0
59    1.0
61    1.0
66    3.0
73    1.0
dtype: float64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [93]:
# Write your code here

import pandas as pd

# Assuming you have a DataFrame named 'df'
# If the DataFrame is not named, replace 'df' with the actual name of your DataFrame

# Create the 'ScoreDifference' column using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Find the index of the game with the highest score difference for each season
max_score_diff_index = df.groupby('Season')['ScoreDifference'].idxmax()

# Display the game with the highest score difference for each season
result_df = df.loc[max_score_diff_index]
print(result_df)

   Season Daynum  Wteam  Wscore  Lteam  Lscore Wloc Numot  ScoreDifference  \
1   49310   1997     66    1157     61    1204    H     0              -47   
3   85042   2005     66    1131     73    1216    H     0              -85   
2   89021   2006     44    1284     41    1343    A     0              -59   
0  100027   2008     66    1203     49    1387    H     0             -184   
4  103660   2009     26    1326     59    1359    H     0              -33   

  HighScoringGame  
1             Yes  
3             Yes  
2             Yes  
0             Yes  
4             Yes  


# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [94]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0    1203    1387
1    1157    1204
2    1284    1343
3    1131    1216
4    1326    1359

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [95]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0    1203    1387
1    1157    1204
2    1284    1343
3    1131    1216
4    1326    1359

Note the difference is the return types when you use brackets and when you use double brackets. 

In [96]:
type(df['Wscore'])

pandas.core.series.Series

In [97]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [98]:
df[0:3]

Season Daynum  Wteam  Wscore  Lteam  Lscore Wloc Numot  ScoreDifference  \
0  100027   2008     66    1203     49    1387    H     0             -184   
1   49310   1997     66    1157     61    1204    H     0              -47   
2   89021   2006     44    1284     41    1343    A     0              -59   

  HighScoringGame  
0             Yes  
1             Yes  
2             Yes

Here's an equivalent using iloc

In [99]:
df.iloc[0:3,:]

Season Daynum  Wteam  Wscore  Lteam  Lscore Wloc Numot  ScoreDifference  \
0  100027   2008     66    1203     49    1387    H     0             -184   
1   49310   1997     66    1157     61    1204    H     0              -47   
2   89021   2006     44    1284     41    1343    A     0              -59   

  HighScoringGame  
0             Yes  
1             Yes  
2             Yes

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [100]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y